[inductor] do mix order reduction fusion earlier and allow fusing more nodes into it

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #168209

This is to handle the pattern we found in internal models. The same weight/bias may be used in multiple layernorm/rmsnorms. That causes mix-order reduction fusing too much stuff together.

Usually for a single rms norm, the backward is like 
- a INNER_RED fused with a PW (pointwise) to compute dx
- a OUTER_RED computing dw

Mix order reduction can fussed all these 3 into a single kernel.

But now, because of sharing weights, INNER_RED_2 from the other norm may get fused in. It would be even worse if there are more than 2 norms sharing weight/bias.

Fusing across different norm backward just saves accessing of weight/bias, it does more harm than benefit.

This PR do a few things to make sure we fuse these nodes properly.
1. give mix order reduction a proper score so we would fuse them before fusions due accessing the shared weight
2. allow fusing extra nodes into a fused-mix-order-reduction.
3. don't fuse extra reduction to a mix order reduction unless the saving is greater or equal to numel of the tensors being reduced. This can skip fusing reductions for another norm backward.

With these change, we can properly generated fused mix-order reduction kernels for each of the norm backward.



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @chenyang78

Differential Revision: [D87548681](https://our.internmc.facebook.com/intern/diff/D87548681)

shunting314 committed:
[inductor] do mix order reduction earlier

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] do mix order reduction earlier"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] do mix order reduction fusion earlier and allow fusing more nodes into it"


This is to handle the pattern we found in internal models. The same weight/bias may be used in multiple layernorm/rmsnorms. That causes mix-order reduction fusing too much stuff together.

Usually for a single rms norm, the backward is like 
- a INNER_RED fused with a PW (pointwise) to compute dx
- a OUTER_RED computing dw

Mix order reduction can fussed all these 3 into a single kernel.

But now, because of sharing weights, INNER_RED_2 from the other norm may get fused in. It would be even worse if there are more than 2 norms sharing weight/bias.

Fusing across different norm backward just saves accessing of weight/bias, it does more harm than benefit.

This PR do a few things to make sure we fuse these nodes properly.
1. give mix order reduction a proper score so we would fuse them before fusions due accessing the shared weight
2. allow fusing extra nodes into a fused-mix-order-reduction.
3. don't fuse extra reduction to a mix order reduction unless the saving is greater or equal to numel of the tensors being reduced. This can skip fusing reductions for another norm backward.

With these change, we can properly generated fused mix-order reduction kernels for each of the norm backward.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben 


[ghstack-poisoned]

PaulZhang12 commented:
@PaulZhang12 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D87548681).

v0i0 commented on torch/_inductor/choices.py:
```diff
@@ -52,6 +53,34 @@ class Sortable(typing.Protocol):
+        if self.template_score != other.template_score:
+            return self.template_score < other.template_score
+
+        if self.memory_score > other.memory_score * threshold:
```
```suggestion
        if min(self.memory_score, other.memory_score * threshold) < max(self.memory_score, other.memory_score):
            return self.memory_score < other.memory_score
```

nit; seems more clear

v0i0 commented on torch/_inductor/scheduler.py:
```diff
@@ -2050,6 +2062,49 @@ def __init__(self, node1: BaseSchedulerNode, node2: BaseSchedulerNode) -> None:
+                self.node2, other
+            )
+        else:
+            return self.sub_node_can_fuse(
```
can this actually happen given that we check is_reduction in sub_node_can_fuse?

eellison commented on torch/_inductor/dependencies.py:
```diff
@@ -339,6 +349,12 @@ def numbytes_hint(self) -> int:
             return 0  # NoneLayout, MultiOutputLayout, etc
 
+    def numel_hint(self) -> int:
+        try:
```
same comment

eellison reviewed (commented):
looks good, have some questions.

shunting314 committed:
Update on "[inductor] do mix order reduction fusion earlier and allow fusing more nodes into it"


This is to handle the pattern we found in internal models. The same weight/bias may be used in multiple layernorm/rmsnorms. That causes mix-order reduction fusing too much stuff together.

Usually for a single rms norm, the backward is like 
- a INNER_RED fused with a PW (pointwise) to compute dx
- a OUTER_RED computing dw

Mix order reduction can fussed all these 3 into a single kernel.

But now, because of sharing weights, INNER_RED_2 from the other norm may get fused in. It would be even worse if there are more than 2 norms sharing weight/bias.

Fusing across different norm backward just saves accessing of weight/bias, it does more harm than benefit.

This PR do a few things to make sure we fuse these nodes properly.
1. give mix order reduction a proper score so we would fuse them before fusions due accessing the shared weight
2. allow fusing extra nodes into a fused-mix-order-reduction.
3. don't fuse extra reduction to a mix order reduction unless the saving is greater or equal to numel of the tensors being reduced. This can skip fusing reductions for another norm backward.

With these change, we can properly generated fused mix-order reduction kernels for each of the norm backward.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

Differential Revision: [D87548681](https://our.internmc.facebook.com/intern/diff/D87548681)

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] do mix order reduction fusion earlier and allow fusing more nodes into it"


This is to handle the pattern we found in internal models. The same weight/bias may be used in multiple layernorm/rmsnorms. That causes mix-order reduction fusing too much stuff together.

Usually for a single rms norm, the backward is like 
- a INNER_RED fused with a PW (pointwise) to compute dx
- a OUTER_RED computing dw

Mix order reduction can fussed all these 3 into a single kernel.

But now, because of sharing weights, INNER_RED_2 from the other norm may get fused in. It would be even worse if there are more than 2 norms sharing weight/bias.

Fusing across different norm backward just saves accessing of weight/bias, it does more harm than benefit.

This PR do a few things to make sure we fuse these nodes properly.
1. give mix order reduction a proper score so we would fuse them before fusions due accessing the shared weight
2. allow fusing extra nodes into a fused-mix-order-reduction.
3. don't fuse extra reduction to a mix order reduction unless the saving is greater or equal to numel of the tensors being reduced. This can skip fusing reductions for another norm backward.

With these change, we can properly generated fused mix-order reduction kernels for each of the norm backward.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

Differential Revision: [D87548681](https://our.internmc.facebook.com/intern/diff/D87548681)

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] do mix order reduction fusion earlier and allow fusing more nodes into it"


This is to handle the pattern we found in internal models. The same weight/bias may be used in multiple layernorm/rmsnorms. That causes mix-order reduction fusing too much stuff together.

Usually for a single rms norm, the backward is like 
- a INNER_RED fused with a PW (pointwise) to compute dx
- a OUTER_RED computing dw

Mix order reduction can fussed all these 3 into a single kernel.

But now, because of sharing weights, INNER_RED_2 from the other norm may get fused in. It would be even worse if there are more than 2 norms sharing weight/bias.

Fusing across different norm backward just saves accessing of weight/bias, it does more harm than benefit.

This PR do a few things to make sure we fuse these nodes properly.
1. give mix order reduction a proper score so we would fuse them before fusions due accessing the shared weight
2. allow fusing extra nodes into a fused-mix-order-reduction.
3. don't fuse extra reduction to a mix order reduction unless the saving is greater or equal to numel of the tensors being reduced. This can skip fusing reductions for another norm backward.

With these change, we can properly generated fused mix-order reduction kernels for each of the norm backward.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

Differential Revision: [D87548681](https://our.internmc.facebook.com/intern/diff/D87548681)

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] do mix order reduction fusion earlier and allow fusing more nodes into it"


This is to handle the pattern we found in internal models. The same weight/bias may be used in multiple layernorm/rmsnorms. That causes mix-order reduction fusing too much stuff together.

Usually for a single rms norm, the backward is like 
- a INNER_RED fused with a PW (pointwise) to compute dx
- a OUTER_RED computing dw

Mix order reduction can fussed all these 3 into a single kernel.

But now, because of sharing weights, INNER_RED_2 from the other norm may get fused in. It would be even worse if there are more than 2 norms sharing weight/bias.

Fusing across different norm backward just saves accessing of weight/bias, it does more harm than benefit.

This PR do a few things to make sure we fuse these nodes properly.
1. give mix order reduction a proper score so we would fuse them before fusions due accessing the shared weight
2. allow fusing extra nodes into a fused-mix-order-reduction.
3. don't fuse extra reduction to a mix order reduction unless the saving is greater or equal to numel of the tensors being reduced. This can skip fusing reductions for another norm backward.

With these change, we can properly generated fused mix-order reduction kernels for each of the norm backward.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

Differential Revision: [D87548681](https://our.internmc.facebook.com/intern/diff/D87548681)

[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge -i

shunting314 commented:
@pytorchbot merge -i
